Reduced amount of ThingOwner patches and total game startup time

Glucocorticoid opened the pull request:
## Changes

Replaced hundreds of generic ThingOwner<T> patches for several ThingOwner patches.

## Reasoning

On heavily modded game CE Controller take a long time (11 seconds) to patch all specific realisations of ThingOwner<T>:
```
11739.3835ms (self: 20.2696 ms) Loading CombatExtended.Controller mod class
- 11719.0974ms (self: 1.1003 ms) Apply Harmony patches
- - 1178.0265ms (self: 1178.0265 ms) instance.PatchAll
- - - 10244.3970ms (self: 4.6578 ms) PatchThingOwner
- - - - 0.0255ms (self: 0.0255 ms) Getting postfix methods
- - - - 94.6166ms (self: 94.6166 ms) Getting patching class collection
- - - - - 23.2577ms (self: 23.2577 ms) Patching Verse.Pawn
- - - - - 16.8603ms (self: 16.8603 ms) Patching Verse.Building
- - - - - 27.9172ms (self: 27.9172 ms) Patching Verse.Corpse
- - - - - 17.9170ms (self: 17.9170 ms) Patching Verse.Explosion
and 460+ patches for Thing subclasses (flames, motes, vanilla tornadoes and so on)
```
10+ seconds for Patch ThingOwner<T>

This patches in result just call compInventory.UpdateInventory() method to calculate total weight and bulk of pawn inventory.
Because of
- ThingOwner<T>.TryAdd method call base.NotifyAdded() and base.NotifyAddedAndMergedWith()
- ThingOwner<T>.Take method call base.Take()
- ThingOwner<T>.Remove method call base.NotifyRemoved()
it is enough to patch those 4 methods of base (non-generic ThingOwner) class for compInventory notification purposes.

In total of this changes CombatExtended.Controller executes much faster:
```
1459.7672ms (self: 11.8943 ms) Loading CombatExtended.Controller mod class
- 1447.8578ms (self: 0.7560 ms) Apply Harmony patches
- - 1140.1545ms (self: 1140.1545 ms) instance.PatchAll
- - - 21.2170ms (self: 0.2953 ms) PatchThingOwner
- - - - 20.9217ms (self: 20.9217 ms) Getting postfix methods
- - - 100.1254ms (self: 100.1254 ms) PatchHediffWithComps
- - - 118.6762ms (self: 118.6762 ms) GenRadial_RadialPatternCount
- - - 21.8618ms (self: 21.8618 ms) PawnColumnWorkers_Resize
- - - 45.0669ms (self: 45.0669 ms) PawnColumnWorkers_SwapButtons
 ```
PatchThingOwner now takes only 21 ms instead of 10200 ms.

Pawns inventory still updates correctly when pawn takes or drops items, but game starts 10+ seconds faster.
More mods player have - more time ThingOwner<T> patches takes.

kbatbouta commented:
Something doesn't make any sense. Where did you change the old patches? You only added new ones and removed annotations that are commented out.

Glucocorticoid commented:
> Something doesn't make any sense. Where did you change the old patches? You only added new ones and removed annotations that are commented out.

Harmony annotations does not needed for manual patching. Because of target methods to patch was changed, signature of patch methods are also changed and does not correspond with old commented Harmony annotations. Name of patch classes also was changed for better matching with target classes. Patches with similar signature was separated between each other for performance analysis purpose to measure TPS consumption independently. If fore some reasons those annotations are strongly needed i will return them.

Key changes are made in PatchThingOwner() method of HarmonyBase class. New patches do their work with parrent class of generic ThingOwner<T>.

kbatbouta commented:
> > Something doesn't make any sense. Where did you change the old patches? You only added new ones and removed annotations that are commented out.
> 
> Harmony annotations does not needed for manual patching. Because of target methods to patch was changed, signature of patch methods are also changed and does not correspond with old commented Harmony annotations. Name of patch classes also was changed for better matching with target classes. Patches with similar signature was separated between each other for performance analysis purpose to measure TPS consumption independently. If for some reasons those annotations are strongly needed i will return them.
> 
> Key changes are made in PatchThingOwner() method of HarmonyBase class. New patches do their work with parrent class of generic ThingOwner<T>.

I'm aware of how annotation and patching works. You misunderstood my question. Where did you remove the old patch? I can't find it in the diff.

Glucocorticoid commented:
> > > Something doesn't make any sense. Where did you change the old patches? You only added new ones and removed annotations that are commented out.
> > 
> > 
> > Harmony annotations does not needed for manual patching. Because of target methods to patch was changed, signature of patch methods are also changed and does not correspond with old commented Harmony annotations. Name of patch classes also was changed for better matching with target classes. Patches with similar signature was separated between each other for performance analysis purpose to measure TPS consumption independently. If for some reasons those annotations are strongly needed i will return them.
> > Key changes are made in PatchThingOwner() method of HarmonyBase class. New patches do their work with parrent class of generic ThingOwner.
> 
> I'm aware of how annotation and patching works. You misunderstood my question. Where did you remove the old patch? I can't find it in the diff.

Oh, i see. Whole old method body after line 65 was commented out by /* */  in 66 and 78 lines.

kbatbouta commented:
> > > > Something doesn't make any sense. Where did you change the old patches? You only added new ones and removed annotations that are commented out.
> > > 
> > > 
> > > Harmony annotations does not needed for manual patching. Because of target methods to patch was changed, signature of patch methods are also changed and does not correspond with old commented Harmony annotations. Name of patch classes also was changed for better matching with target classes. Patches with similar signature was separated between each other for performance analysis purpose to measure TPS consumption independently. If for some reasons those annotations are strongly needed i will return them.
> > > Key changes are made in PatchThingOwner() method of HarmonyBase class. New patches do their work with parrent class of generic ThingOwner.
> > 
> > 
> > I'm aware of how annotation and patching works. You misunderstood my question. Where did you remove the old patch? I can't find it in the diff.
> 
> Oh, i see. Whole old method body after line 65 was commented out by /* */ in 66 and 78 lines.

Can verify that `ThingOwner<T>` call these methods on every change? If the answer is yes then i might request some things that could speed up the inventory system by a lot. (The last part is optional)

MadaraUchiha reviewed (dismissed):
LGTM! Small nitpick

Glucocorticoid commented:
> Can verify that ThingOwner<T> call these methods on every change? If the answer is yes then i might request some things that could speed up the inventory system by a lot. (The last part is optional)
> 
> If you can verify caravans and things alike to be working that would be great!

Well.  I am check latest beta version of RW assembly in steam for 1.3 and latest version of 1.2. They both has same code of ThingOwner classes. 
ThingOwner<T>.Remove return false if ThingOwner.Contains(item) return false (owner does not have an item) or call ThingOwner.NotifyRemoved(item) before exit. There is no way to miss new patch.
ThingOwner<T>.TryAdd(Thing, bool) has a lot of negative exits when item does not added and two positive exits with ThingOwner.NotifyAddedAndMergedWith() and ThingOwner.base.NotifyAdded() calls. New patches are on board.
ThingOwner<T>.Take methods calls non generic versions of base class without any conditions.

I made this patches 3 days ago on local version of CE and done a lot of tests. Does not found any bugs. 

> Please delete unneeded code instead of commenting it out
Deleted.

MadaraUchiha reviewed (approved):
@CombatExtended-Continued/playtesters please add your approval before merge :)

Alicecomma commented:
LGTM, glad there's a better way to do it now!
